File: backend/query/summarizer.py
from clients.gemini import client
from query.retrieval import get_all_chunks
from utils.ai_handler import generate_with_fallback


def _summarize_chunk(chunk_text: str, index: int) -> str:
    map_prompt = f"""
                    You are an expert content summarizer. Your task is to summerise given transcript section in concise and to the point manner.
                    Summarize the following section (part {index + 1}) of a  video transcript.
                    Include the key points and topics discussed.
                    Keep the summary concise and to the point.
                    Maintain original tone and style of the transcript.
                    Do not add any extra information or commentary.


                    Transcript Section:
                    {chunk_text}
                """
    return generate_with_fallback(
        client=client,
        prompt=map_prompt,
        primary_model="gemini-3.1-flash-lite-preview",
        fallback_model="gemini-2.5-flash",
        config={"temperature": 0.1},
    )


def map_reduce_summary(video_id: str) -> str:
    chunks = get_all_chunks(video_id)
    if not chunks:
        return "No indexed content available to summarize. Please process the video first."

    chunk_summaries = [_summarize_chunk(text, i) for i, text in enumerate(chunks)]

    combined_summaries = "\n\n--- Next Section ---\n\n".join(chunk_summaries)
    reduce_prompt = f"""
                        You are provided with chronological summaries of different sections of a video.
                        Create a comprehensive, cohesive, and structured final summary of the entire video.
                        Use headings and bullet points where appropriate.
                        Do not add any extra information or commentary.
                        Maintain original tone and style of the transcript.
                        

                        Chronological Segment Summaries:
                        {combined_summaries}
                    """
    return generate_with_fallback(
        client=client,
        prompt=reduce_prompt,
        primary_model="gemini-3.1-flash-lite-preview",
        fallback_model="gemini-2.5-flash",
        config={"temperature": 0.1},
    )

File: frontend/src/lib/rag/query/summarizer.ts
import { generateWithFallback, streamWithFallback } from "../ai-handler";
import { CHAT_MODEL_FAST, CHAT_MODEL_STRONG, SUMMARY_MAP_CONCURRENCY } from "../constants";
import { getAllChunks } from "./retrieval";

async function summarizeChunk(chunkText: string, index: number): Promise<string> {
  const mapPrompt = `
Summarize the following section (part ${index + 1}) of a video transcript.
Include the key points and topics discussed. Keep it concise.
Do not add extra information.

Transcript Section:
${chunkText}
`;

  return generateWithFallback(mapPrompt, CHAT_MODEL_FAST, CHAT_MODEL_STRONG, 0.1);
}

export async function mapReduceSummary(
  videoId: string,
  language: string,
  onToken?: (token: string) => void,
): Promise<string> {
  const chunks = await getAllChunks(videoId);
  if (!chunks.length) {
    return "No indexed content available to summarize. Please process the video first.";
  }

  const chunkSummaries: string[] = [];

  for (let i = 0; i < chunks.length; i += SUMMARY_MAP_CONCURRENCY) {
    const batch = chunks.slice(i, i + SUMMARY_MAP_CONCURRENCY);
    const batchSummaries = await Promise.all(
      batch.map((chunkText, j) => summarizeChunk(chunkText, i + j)),
    );
    chunkSummaries.push(...batchSummaries);
  }

  const combinedSummaries = chunkSummaries.join("\n\n--- Next Section ---\n\n");
  const reducePrompt = `
You are provided with chronological summaries of different sections of a video.
Create a comprehensive, cohesive, and structured final summary of the entire video.
Use headings and bullet points where appropriate.
Do not add extra information.

Write the entire summary in ${language} only — even if the transcript segments are in another language.

Chronological Segment Summaries:
${combinedSummaries}
`;

  if (onToken) {
    return streamWithFallback(
      reducePrompt,
      CHAT_MODEL_FAST,
      CHAT_MODEL_STRONG,
      onToken,
      0.1,
    );
  }

  return generateWithFallback(reducePrompt, CHAT_MODEL_FAST, CHAT_MODEL_STRONG, 0.1);
}
